helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkEnvironment.java

```diff
@@ -0,0 +1,448 @@
+
+  private MockMetadataStoreDirectoryServer _msds;
+  private static final Map<String, Collection<String>> _rawRoutingData = new HashMap<>();
+  private RealmAwareZkClient _zkClient;
+  private HelixAdmin _zkHelixAdmin;
+
+  @BeforeClass
+  public void beforeClass() throws Exception {
+    // Create 3 in-memory zookeepers and routing mapping
+    final String zkPrefix = "localhost:";
+    final int zkStartPort = 8777;
+
+    for (int i = 0; i < NUM_ZK; i++) {
+      String zkAddress = zkPrefix + (zkStartPort + i);
+      ZK_SERVER_MAP.put(zkAddress, TestHelper.startZkServer(zkAddress));
+      ZK_CLIENT_MAP.put(zkAddress, DedicatedZkClientFactory.getInstance()
+          .buildZkClient(new HelixZkClient.ZkConnectionConfig(zkAddress),
+              new HelixZkClient.ZkClientConfig().setZkSerializer(new ZNRecordSerializer())));
+
+      // One cluster per ZkServer created
+      _rawRoutingData.put(zkAddress, Collections.singletonList("/" + CLUSTER_LIST.get(i)));
+    }
+
+    // Create a Mock MSDS
+    final String msdsHostName = "localhost";
+    final int msdsPort = 11117;
+    final String msdsNamespace = "multiZkTest";
+    _msds = new MockMetadataStoreDirectoryServer(msdsHostName, msdsPort, msdsNamespace,
+        _rawRoutingData);
+    _msds.startServer();
+
+    // Turn on multiZk mode in System config
+    System.setProperty(SystemPropertyKeys.MULTI_ZK_ENABLED, "true");
+    // MSDS endpoint: http://localhost:11117/admin/v2/namespaces/multiZkTest
+    System.setProperty(MetadataStoreRoutingConstants.MSDS_SERVER_ENDPOINT_KEY,
+        "http://" + msdsHostName + ":" + msdsPort + "/admin/v2/namespaces/" + msdsNamespace);
+
+    // Create a FederatedZkClient for admin work
+    _zkClient =
+        new FederatedZkClient(new RealmAwareZkClient.RealmAwareZkConnectionConfig.Builder().build(),
+            new RealmAwareZkClient.RealmAwareZkClientConfig());
+  }
+
+  @AfterClass
+  public void afterClass() throws Exception {
+    // Kill all mock controllers and participants
+    MOCK_CONTROLLERS.values().forEach(ClusterControllerManager::syncStop);
+    MOCK_PARTICIPANTS.forEach(MockParticipantManager::syncStop);
+
+    // Tear down all clusters
+    CLUSTER_LIST.forEach(cluster -> TestHelper.dropCluster(cluster, _zkClient));
+
+    // Verify that all clusters are gone in each zookeeper
+    Assert.assertTrue(TestHelper.verify(() -> {
+      for (Map.Entry<String, HelixZkClient> zkClientEntry : ZK_CLIENT_MAP.entrySet()) {
+        List<String> children = zkClientEntry.getValue().getChildren("/");
+        if (children.stream().anyMatch(CLUSTER_LIST::contains)) {
+          return false;
+        }
+      }
+      return true;
+    }, TestHelper.WAIT_DURATION));
+
+    // Tear down zookeepers
+    ZK_SERVER_MAP.forEach((zkAddress, zkServer) -> zkServer.shutdown());
+
+    // Stop MockMSDS
+    _msds.stopServer();
+
+    // Turn off multiZK mode in System config
+    System.clearProperty(SystemPropertyKeys.MULTI_ZK_ENABLED);
```

Comment:
Just a corner case, if there is exception before clearProperty(), this property will be kept and may affect other tests? Usually what I do is using try...finally.

Reply:
Again, I can wrap it in try-finally but each test is responsible for controlling its own System property configs. So a well-written test with a proper beforeClass or a setup method shouldn't be affected. But I agree - wrapping it in a try-finally is not a bad idea so I will add it.